### PR TITLE
consensus/pbft: check proposer in core engine

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/pbft/backend.go [hunk 1]
```diff
@@ -41,8 +41,8 @@ type Backend interface {
 	// Commit is to deliver a final result to write into blockchain
 	Commit(Proposal) error
 
-	// RoundChanged is called when round change occurred
-	RoundChanged(needNewProposal bool) error
+	// NextSeal is called when we want to trigger next Seal()
+	NextSeal() error
```
Comment (markya0616) [inline consensus/pbft/backend.go]: NextRound

File: consensus/pbft/core/backlog.go [hunk 2]
```diff
@@ -31,30 +31,35 @@ var (
 	}
 )
 
-// check whether it's a future message
-// It's a future message if the message priority is smaller than current priority
-func (c *core) isFutureMessage(msgCode uint64, view *pbft.View) bool {
+// check the message state
+// return errInvalidMessage if the message is invalid
+// return errFutureMessage if the message priority is smaller than current priority
```
Comment (tailingchen) [inline consensus/pbft/core/backlog.go]: how about replace `message priority` with `message view`?

File: consensus/pbft/core/core.go [hunk 3]
```diff
@@ -223,6 +228,9 @@ func (c *core) setState(state State) {
 		c.state = state
 	}
 	c.processBacklog()
+	if state == StateAcceptRequest {
+		c.processPendingRequests()
```
Comment (tailingchen) [inline consensus/pbft/core/core.go]: `c.processPendingRequests()` should be processed before `c.processBacklog()`.

File: consensus/pbft/backends/simple/backend.go [hunk 4]
```diff
@@ -145,15 +145,15 @@ func (sb *simpleBackend) Commit(proposal pbft.Proposal) error {
...
-func (sb *simpleBackend) RoundChanged(needNewProposal bool) error {
-	if needNewProposal {
-		// if I proposed a block before, fire a ChainHeadEvent to trigger next Seal()
-		// and the previous Seal() will be stopped.
-		if !common.EmptyHash(sb.proposedBlockHash) {
-			go sb.eventMux.Post(core.ChainHeadEvent{})
-		}
+// NextSeal will broadcast ChainHeadEvent to trigger next seal()
+func (sb *simpleBackend) NextSeal() error {
+	if sb.chain == nil {
```
Comment (bailantaotao) [inline consensus/pbft/backends/simple/backend.go]: 如果不會發生的話就刪掉他吧

File: consensus/pbft/core/request.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ import (
 
 func (c *core) handleRequest(request *pbft.Request, p pbft.Validator) error {
 	logger := c.logger.New("state", c.state)
```
Comment (bailantaotao) [inline consensus/pbft/core/request.go]: 記錄更多的訊息？例如來源地址
```
c.logger.New("from", src.Address().Hex(), "state", c.state)
```

File: consensus/pbft/config.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ type Config struct {
 var DefaultConfig = &Config{
 	RequestTimeoutMsec: 10000,
 	BlockPeriod:        1,
-	BlockPauseTime:     2,
+	BlockPauseTime:     1,
```
Comment (bailantaotao) [inline consensus/pbft/config.go]: 為什麼改成1?不是說要大於BlockPeriod?

File: consensus/pbft/core/backlog.go (hunk 2 above)
Comment (markya0616) [inline consensus/pbft/core/backlog.go]: OK

File: consensus/pbft/core/core.go (hunk 3 above)
Comment (markya0616) [inline consensus/pbft/core/core.go]: OK

File: consensus/pbft/backends/simple/backend.go (hunk 4 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: OK

File: consensus/pbft/core/request.go (hunk 5 above)
Comment (markya0616) [inline consensus/pbft/core/request.go]: 因為request 都會從自己來  所以我就不特別印了

File: consensus/pbft/config.go (hunk 6 above)
Comment (markya0616) [inline consensus/pbft/config.go]: 漚漚漚  我只是要方便測試XD  把它改回來

File: consensus/pbft/backends/simple/backend.go [hunk 7]
```diff
@@ -145,14 +145,10 @@ func (sb *simpleBackend) Commit(proposal pbft.Proposal) error {
...
-	if sb.chain == nil {
-		sb.logger.Info("NextSeal", "address", sb.Address().Hex())
-	} else {
-		header := sb.chain.CurrentHeader()
-		sb.logger.Info("NextSeal", "address", sb.Address().Hex(), "current_hash", header.Hash(), "current_number", header.Number)
-	}
+// NextRound will broadcast ChainHeadEvent to trigger next seal()
+func (sb *simpleBackend) NextRound() error {
+	header := sb.chain.CurrentHeader()
+	sb.logger.Info("NextRound", "address", sb.Address().Hex(), "current_hash", header.Hash(), "current_number", header.Number)
```
Comment (alanchchen) [inline consensus/pbft/backends/simple/backend.go]: I suggest using `sb.logger.Debug` or `sb.logger.Trace` here. It's not an important function.

File: consensus/pbft/backends/simple/engine_test.go [hunk 8]
```diff
@@ -166,7 +166,7 @@ func TestSealRoundChange(t *testing.T) {
 			if !ok {
 				t.Errorf("unexpected event comes, got: %v, expected: pbft.RequestEvent", reflect.TypeOf(ev.Data))
 			}
-			engine.RoundChanged(true)
+			engine.NextSeal()
```
Comment (alanchchen) [inline consensus/pbft/backends/simple/engine_test.go]: NextRound ?

File: consensus/pbft/backends/simple/backend.go (hunk 7 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: OK

File: consensus/pbft/backends/simple/engine_test.go (hunk 8 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/engine_test.go]: OK

**Pull request merged**